addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,327 @@
+                    }
+                } else if (command.equals("dimmer")) {
+                    updateState(getThing().getChannel("dimmer").getUID(), readDimmer());
+                } else if (command.equals("freq")) {
+                    updateState(getThing().getChannel("frequency").getUID(), readFrequency());
+                } else if (command.equals("source")) {
+                    updateState(getThing().getChannel("source").getUID(), new StringType(readUntil('!')));
+                } else if (command.equals("display")) {
+                    String str_length = readUntil(',');
+                    int length = Integer.parseInt(str_length);
+                    byte[] data = new byte[length];
+                    int off = 0;
+                    while (off != length) {
+                        int r = serialPort.getInputStream().read(data, off, length - off);
+                        if (r == -1) {
+                            throw new IOException("Connection closed while reading display content");
+                        } else {
+                            off += r;
+                        }
+                    }
+                    // We don't do anything with display content, could add a channel if this is useful
+                } else {
+                    readUntil('!'); // discard
+                }
+
+            } catch (IOException e) {
+                logger.error("Input error while receiving data from amplifier", e);
+                disconnect();
+            }
+        }
+    }
+
+    void send(String text) throws IOException {
+        if (power) {
+            connect();
+            serialPort.getOutputStream().write(text.getBytes());
+        }
+    }
+
+    void sendForce(String text) throws IOException {
+        connect();
+        serialPort.getOutputStream().write(text.getBytes());
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        try {
+            if (channelUID.getId().equals("power")) {
+                if (command == OnOffType.ON) {
+                    sendForce("power_on!");
+                } else if (command == OnOffType.OFF) {
+                    sendForce("power_off!");
+                } else if (command instanceof RefreshType) {
+                    sendForce("get_current_power!");
+                }
+            } else if (channelUID.getId().equals("mute")) {
+                if (command == OnOffType.ON) {
+                    send("mute_on!");
+                } else {
+                    send("mute_off!");
+                }
+            } else if (channelUID.getId().equals("volume")) {
+                handleVolume(command);
+            } else if (channelUID.getId().equals("dimmer")) {
+                // Invert the scale so 100% is brightest
+                if (command instanceof PercentType) {
+                    double value = 6 - Math.floor(((PercentType) command).doubleValue() * 6 / 100.0);
```

Comment:
consider to handle this in an external method like you did with `handleVolume`, optionally do that for all bodies.